csharp/src/Ice/Properties.cs

```diff
@@ -164,17 +454,371 @@ public interface Properties
+        int dotPos = key.IndexOf('.');
+
+        // If the key doesn't contain a dot, it's not a valid Ice property
+        if (dotPos == -1)
+        {
+            return null;
+        }
+
+        string prefix = key.Substring(0, dotPos);
+
+        Property[]? propertyArray = null;
+
+        // Search for the property list that matches the prefix
+        foreach (Property[]? validProps in PropertyNames.validProps)
+        {
+            string pattern = validProps[0].pattern;
+            dotPos = pattern.IndexOf('.');
+
+            // Each top level prefix describes a non-empty
+            // namespace. Having a string without a prefix followed by a
+            // dot is an error.
+            Debug.Assert(dotPos != -1);
+
+            // Trim any leading/trailing ^, $, or \ characters from the prefix
+            string propPrefix = pattern.Substring(0, dotPos).TrimStart(['^', '$', '\\']);
+
+            if (propPrefix == prefix)
+            {
+                // We found the property list that matches the prefix
+                propertyArray = validProps;
+                break;
+            }
+
+            // As a courtesy to the user, perform a case-insensitive match and suggest the correct property.
+            // Otherwise no other warning is issued.
+            if (logWarnings && propPrefix.ToUpper().Equals(prefix.ToUpper()))
+            {
+                logger.warning("unknown property: `" + key + "'; did you mean `" + propPrefix + "'?");
+                return null;
+            }
+        }
+
+        if (propertyArray == null)
+        {
+            // The prefix is not a valid Ice property
+            return null;
+        }
+
+        foreach (var prop in propertyArray)
+        {
+            bool matches = prop.usesRegex ? Regex.IsMatch(key, prop.pattern) : key == prop.pattern;
+            if (matches)
+            {
+                if (prop.deprecated && logWarnings)
+                {
+                    logger.warning("deprecated property: " + key);
+                }
+                return prop;
+            }
+        }
+
+        // If we get here, the prefix is valid but the property is unknown
+        if (logWarnings)
+        {
+            logger.warning("unknown property: " + key);
+        }
+        return null;
+    }
+
+    /// <summary>
+    /// Gets the default value for a given Ice property.
+    /// </summary>
+    /// <param name="key">The Ice property key</param>
+    /// <returns>The default property value, or an empty string the default is unspecified.</returns>
+    /// <exception cref="ArgumentException"></exception>
+    private static string getDefaultProperty(string key)
+    {
+        // Find the property, don't log any warnings.
+        Property? prop = findProperty(key, false) ?? throw new ArgumentException("unknown ice property: " + key);
+        return prop.defaultValue;
+    }
+
+    private Dictionary<string, PropertyValue> _properties;
```

Comment:
I would initialize the properties here
```suggestion
    private readonly Dictionary<string, PropertyValue> _properties = [];
```

And no need to the same in both constructors.